gazebo_plugins/src/gazebo_ros_imu_sensor.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2018 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <gazebo_plugins/gazebo_ros_imu_sensor.hpp>
+#include <gazebo_ros/conversions.hpp>
+#include <gazebo_ros/utils.hpp>
+
+#include <iostream>
+#include <memory>
+#include <string>
+
+namespace gazebo_plugins
+{
+
+class GazeboRosImuSensorPrivate
+{
+public:
+  /// Node for ros communication
+  gazebo_ros::Node::SharedPtr rosnode_;
+  /// Publish for imu message
+  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr pub_;
+  /// IMU message modified each update
+  sensor_msgs::msg::Imu::SharedPtr msg_;
+  /// IMU sensor this plugin is attached to
+  gazebo::sensors::ImuSensorPtr sensor_;
+  /// Event triggered when sensor updates
+  gazebo::event::ConnectionPtr sensor_update_event_;
+
+  /// Publish latest imu data to ROS
+  void OnUpdate();
+};
+
+GazeboRosImuSensor::GazeboRosImuSensor()
+: impl_(std::make_unique<GazeboRosImuSensorPrivate>())
+{
+}
+
+GazeboRosImuSensor::~GazeboRosImuSensor()
+{
+}
+
+void GazeboRosImuSensor::Load(gazebo::sensors::SensorPtr _sensor, sdf::ElementPtr _sdf)
+{
+  impl_->rosnode_ = gazebo_ros::Node::Create("gazebo_ros_imu_sensor", _sdf);
+
+  impl_->sensor_ = std::dynamic_pointer_cast<gazebo::sensors::ImuSensor>(_sensor);
+  if (!impl_->sensor_) {
+    RCLCPP_ERROR(impl_->rosnode_->get_logger(), "Parent is not an imu sensor. Exiting.");
+    return;
+  }
+
+  impl_->pub_ = impl_->rosnode_->create_publisher<sensor_msgs::msg::Imu>("~/out");
+
+  // Create message to be reused
+  auto msg = std::make_shared<sensor_msgs::msg::Imu>();
```

Comment:
@mikaelarguedas can you confirm that this will initialise the other elements of the covariance matrix fields to 0?
From https://github.com/ros2/geometry2/pull/64 I think the answer is yes, but http://design.ros2.org/articles/generated_interfaces_cpp.html says the default constructor causes default/value initialised, and doesn't say the default is zero initialised.

Reply:
TL;DR: yes all the fields of the array will be set to zero

Longer answer:
From the doc: `The default constructor initializes all members with their default value; if a field doesn’t have a default value, then the member is value-initialized.`

As Imu doesn't provide custom default values for its' fields (https://github.com/ros2/common_interfaces/blob/60afcc4b3256d5a0f606939a5d02e173ff15bdd9/sensor_msgs/msg/Imu.msg#L18) they will initialized with ROS' default.

These default are language independent so they are defined in the [Interface Definition](http://design.ros2.org/articles/interface_definition.html) doc.

```
A field can optionally specify a default value. If no default value is specified a common default value is used:

for bool it is false
for numeric types it is a 0 value
for string it is an empty string
for static size arrays it is an array of N elements with its fields zero-initialized
for bounded size arrays and dynamic size arrays it is an empty array []
```

The part we're interested in here is `for static size arrays it is an array of N elements with its fields zero-initialized`

Reply:
mmmm I didn't realise from the design article that it would use a ROS default and not just a default specified in the .msg. I will try to update that design doc based on your answer, thanks!!